### PR TITLE
AcraTranslator demo Go server + MongoDB

### PR DESCRIPTION
**Pull request by Zhaars**
A working version of Go API server with MongoDB and AcraTranslator via http-api;

### CODE REVIEW

File: acra-translator/docker-compose.acra-translator.yml [hunk 1]
```diff
@@ -0,0 +1,145 @@
...
+      - ME_CONFIG_MONGODB_AUTH_USERNAME=${MONGO_ROOT_USER:-root}
+      - ME_CONFIG_MONGODB_AUTH_PASSWORD=${MONGO_ROOT_PASSWORD:-password}
+      - ME_CONFIG_BASICAUTH_USERNAME=${MONGOEXPRESS_LOGIN:-test}
+      - ME_CONFIG_BASICAUTH_PASSWORD=${MONGOEXPRESS_PASSWORD:-test}
+    depends_on:
+      - mongo
+    ports:
+      - "8081:8081"
+    networks:
+      - world
```
Comment (Lagovas) [inline acra-translator/docker-compose.acra-translator.yml:63]: why it works without `webui-mongo` network ?)
as I understand, it should render some UI but have problems with connectivity to mongo server

File: acra-translator/docker-compose.acra-translator.yml [hunk 2]
```diff
@@ -0,0 +1,145 @@
...
+    # Restart server after correct termination, for example after the config
+    # was changed through the API
+    restart: always
+    container_name: acra-server
+    depends_on:
+      - acra-keymaker_client
+    environment:
+      ACRA_MASTER_KEY: ${ACRA_SERVER_MASTER_KEY:-N0MyZEVCRGY1cHA4ODRSVHp3c3BNeGtDekxaV3BhTHI=}
+    # We use internal networks:
+    # - 'server-mysql' - for AcraServer and DB interconnection
```
Comment (Lagovas) [inline acra-translator/docker-compose.acra-translator.yml:76]: let's remove redundant comments

File: acra-translator/docker-compose.acra-translator.yml [hunk 3]
```diff
@@ -0,0 +1,145 @@
...
+      - "8081:8081"
+    networks:
+      - world
+
+  acra-translator:
+    image: "cossacklabs/acra-translator:${ACRA_DOCKER_IMAGE_TAG:-0.95.0}"
+    # Restart server after correct termination, for example after the config
+    # was changed through the API
+    restart: always
+    container_name: acra-server
```
Comment (Lagovas) [inline acra-translator/docker-compose.acra-translator.yml:70]: why not name it `acra-translator`?

File: acra-translator/docker-compose.acra-translator.yml [hunk 4]
```diff
@@ -0,0 +1,145 @@
...
+        VCS_URL: ${COSSACKLABS_ACRA_VCS_URL}
+        VCS_REF: ${COSSACKLABS_ACRA_VCS_REF}
+        VCS_BRANCH: ${COSSACKLABS_ACRA_VCS_BRANCH}
+        BUILD_DATE: ${COSSACKLABS_ACRAENGDEMO_BUILD_DATE}
+    environment:
+      ADDRESS: 0.0.0.0:8008
+      SWAGGER_HOST: localhost:8008
+      LOG_LEVEL: debug
+      MONGODB_CONNECTION_URL: mongodb://root:password@mongo:27017/admin?authSource=admin
+      ACRA_TRANSLATOR_URL: https://acra-server:9595
```
Comment (Lagovas) [inline acra-translator/docker-compose.acra-translator.yml:116]: looks confusing why we use `acra-server` as domain name

File: acra-translator/go-translator-demo/README.md [hunk 5]
```diff
@@ -0,0 +1,24 @@
+# cossacklabs go-translator-demo
+
+### Pre requirements
+
+    go version go1.18.0 or higher 
+    docker version 19.03.8
+    swagg # for swagger generatetion
```
Comment (Lagovas) [inline acra-translator/go-translator-demo/README.md:7]:
```suggestion
    swagg # for swagger generation
```

File: acra-translator/go-translator-demo/README.md [hunk 6]
```diff
@@ -0,0 +1,24 @@
...
+
+### Swagger
+
+Swagger is automatically generated from comments in source code using `swag`. To install `swag` locally:
+
+```
+go get -u github.com/swaggo/swag/cmd/swag
+
+# 1.18 or newer
+go install github.com/swaggo/swag/cmd/swag@latest
```
Comment (Lagovas) [inline acra-translator/go-translator-demo/README.md:17]: let's freeze the swag version that was used in tests. Now software updates so frequently that we will have down examples in 3 weeks )

File: acra-translator/go-translator-demo/docker/go-translator-demo.Dockerfile [hunk 7]
```diff
@@ -0,0 +1,82 @@
...
+ARG VCS_REF
+# Repository branch
+ARG VCS_BRANCH
+# Date of the build
+ARG BUILD_DATE
+
+# Include metadata, additionally use label-schema namespace
+LABEL org.label-schema.schema-version="1.0" \
+    org.label-schema.vendor="Cossack Labs" \
+    org.label-schema.url="https://cossacklabs.com" \
```
Comment (Lagovas) [inline acra-translator/go-translator-demo/docker/go-translator-demo.Dockerfile:19]:
```suggestion
    org.label-schema.url="https://www.cossacklabs.com" \
```

File: acra-translator/go-translator-demo/cmd/api/.env.example [hunk 8]
```diff
@@ -0,0 +1,10 @@
+ADDRESS=0.0.0.0:8008
```
Comment (Lagovas) [inline acra-translator/go-translator-demo/cmd/api/.env.example:1]: can you explain where we use this file or how it should be used? I don't see usage in the dockerfile or mounting in docker-compose files, and don't see mention in the readme

File: acra-translator/go-translator-demo/README.md [hunk 9]
```diff
@@ -0,0 +1,41 @@
...
+
+Make sure you install all required dependencies and set up project and env variables.
+
+```
+ADDRESS=0.0.0.0:8008;
+SWAGGER_HOST=0.0.0.0:8008;
+
+ACRA_TRANSLATOR_URL=https://localhost:9595;
+
+# requeied for connection to AcraTranslator
```
Comment (Lagovas) [inline acra-translator/go-translator-demo/README.md:19]:
```suggestion
# required for connection to AcraTranslator
```

File: acra-translator/go-translator-demo/README.md [hunk 10]
```diff
@@ -0,0 +1,41 @@
...
+TLS_CLIENT_CA_PATH=<path to ca.crt>
+TLS_CLIENT_CERT_PATH=<path to client.crt>
+# requeied for connection to AcraTranslator
```
Comment (Lagovas) [inline acra-translator/go-translator-demo/README.md:22]:
```suggestion
# required for connection to AcraTranslator
```

File: acra-translator/go-translator-demo/handlers/user.go [hunk 11]
```diff
@@ -0,0 +1,431 @@
...
+		return
+	}
+
+	if err := user.Validate(); err != nil {
+		render.Error(w, r, errors.NewInvalidInputError(fmt.Sprintf("invalid input provided - %s", err.Error())))
+		return
+	}
+
+	// Here is the example if synchronous call to AcraTranslator.
+	// It can be simplified via using the BulkAPI supported by AcraTranslator Enterprise version
```
Comment (Lagovas) [inline acra-translator/go-translator-demo/handlers/user.go:63]: let's leave url for further reading - https://docs.cossacklabs.com/acra/guides/integrating-acra-translator-into-new-infrastructure/http_api/#bulk-processing-api-enterprise

File: run.sh [hunk 12]
```diff
@@ -307,6 +307,30 @@ Resources that will become available after launch:
...
+Resources that will become available after launch:
+
+    * Container with environment prepared for the AcraTransaltor example.
+
+    * Web interface for MongoDB - see how the encrypted data is stored:
+        http://$HOST:8081
+        BasicAuth user/password: test/test
+
+    * MongoDB - also you can connect to DB directly:
+        mysql://$HOST:27017
```
Comment (Lagovas) [inline run.sh:321]: mysql?)

**Comment by Lagovas** [comment]
in general looks easier to read and understand. but we should describe how to use, and probably add readme for this app with short description of main parts and where to find wrapper around storage, for translator, where communication with storage and web frontend

File: README.md [hunk 13]
```diff
@@ -1342,6 +1342,220 @@ Let's see how many code lines are necessary to encrypt and search some data usin
...
+
+```bash
+$ curl https://raw.githubusercontent.com/cossacklabs/acra-engineering-demo/master/run.sh | \
+    bash -s -- acra-translator
+```
+
+This command downloads a simple Go API Server application that stores the data in a database, AcraTranslator Docker
+containers, MongoDB
+database, sets up the environment, configures Go application to connect to AcraTranslator and MongoDB, and provides an
+API(Swagger) for you to try.
```
Comment (Lagovas) [inline README.md:1362]: can we generate postman config for that? a lot of people relies on that and looks very useful

**Comment by vixentael** [comment]
Niiiice!

File: README.md [hunk 14]
```diff
@@ -1342,6 +1343,221 @@ Let's see how many code lines are necessary to encrypt and search some data usin
...
+
+## 1. Installation
+
+```bash
+$ curl https://raw.githubusercontent.com/cossacklabs/acra-engineering-demo/master/run.sh | \
+    bash -s -- acra-translator
+```
+
+This command downloads a simple Go API Server application, AcraTranslator, MongoDB, sets up the environment, configures
+Go application to data with AcraTranslator and store to MongoDB, and provides an
```
Comment (Lagovas) [inline README.md:1361]:
```suggestion
Go application to encrypt data with AcraTranslator and store to MongoDB, and provides an
```

**Review by Lagovas — APPROVED** [review]
last missing word and let's merge it, great job

**Pull request merged**